src/LanguageServer/Protocol/Features/Options/SolutionAnalyzerConfigOptionsUpdater.cs

```diff
@@ -0,0 +1,273 @@
+                {
+                    case WorkspaceChangeKind.SolutionAdded:
+                    case WorkspaceChangeKind.SolutionChanged:
+                    case WorkspaceChangeKind.SolutionCleared:
+                    case WorkspaceChangeKind.SolutionReloaded:
+                    case WorkspaceChangeKind.SolutionRemoved:
+                    case WorkspaceChangeKind.ProjectAdded:
+                    case WorkspaceChangeKind.ProjectChanged:
+                    case WorkspaceChangeKind.ProjectReloaded:
+                    case WorkspaceChangeKind.ProjectRemoved:
+                        break;
+
+                    default:
+                        return;
+                }
+
+                if (args.OldSolution.SolutionState.ProjectCountByLanguage.KeysEqual(args.NewSolution.SolutionState.ProjectCountByLanguage))
+                {
+                    return;
+                }
+
+                UpdateSolution(
+                    transformation: solution => InitializeLanguages(solution, _initializedLanguages),
+                    onAfterUpdate: solution => _initializedLanguages = solution.SolutionState.ProjectCountByLanguage);
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagate(e, ErrorSeverity.Diagnostic))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+        }
+
+        public void GlobalOptionsChanged(object? sender, OptionChangedEventArgs args)
+        {
+            try
+            {
+                if (!args.ChangedOptions.Any(static o => o.key.Option.Definition.IsEditorConfigOption))
+                {
+                    return;
+                }
+
+                UpdateSolution(
+                    transformation: solution => UpdateOptions(solution, args, _initializedLanguages),
+                    onAfterUpdate: null);
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagate(e, ErrorSeverity.Diagnostic))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+        }
+
+        private void UpdateSolution(Func<Solution, Solution> transformation, Action<Solution>? onAfterUpdate)
+        {
+            var lockTaken = false;
+            try
+            {
+                // If another update is in progress wait until it completes.
+                Monitor.Enter(_updateLock, ref lockTaken);
+
+                workspace.SetCurrentSolution(
+                    transformation,
+                    changeKind: WorkspaceChangeKind.SolutionChanged,
+                    onAfterUpdate: (_, newSolution) =>
+                    {
+                        onAfterUpdate?.Invoke(newSolution);
+
+                        // unlock before workspace events are triggered:
+                        if (lockTaken)
+                        {
+                            Monitor.Exit(_updateLock);
+                            lockTaken = false;
+                        }
+                    });
```

Comment:
really would like it if we could adjust this to use the result of SetCurrentSOlution as well, and then have simple locking/unlocking semantics.

Reply:
How would we do that? We need to unlock before SetCurrentSolution triggers the workspace changed event.

Reply:
unclear to me from reading the code why we care about hte workspace changed event if we can just look at teh final solution created.  i'm likely missing something about this.